pyfr/integrators/multipmeta.py

```diff
@@ -0,0 +1,151 @@
+
+import re
+
+from pyfr.inifile import Inifile
+from pyfr.util import proxylist
+
+
+class MultiPMeta(type):
+    def __new__(mcs, name, bases, attrs):
+        # Initialise new mgintegrator class
+        mgintegrator = super().__new__(mcs, name, bases, attrs)
+
+        # Get config file
+        cfg = attrs['cfg']
+
+        dtau = cfg.getfloat('solver-time-integrator', 'pseudo-dt')
+        order = cfg.getint('solver', 'order')
+
+        # Get multigrid cycle parameters
+        sect = 'pseudo-multigrid'
+        lvliters = list(cfg.getliteral(sect, 'pseudo-mgcycle'))
+        nlevels = len(lvliters)
+
+        if nlevels > order:
+            raise ValueError('The number of multigrid levels cannot exceed '
+                             'the solution order')
+
+        dtaufact = cfg.getfloat(sect, 'pseudo-dt-fact', 1.0)
+        dtaus = [dtaufact**i * dtau for i in range(nlevels)]
+
+        # The maximum and minimum number of multigrid cycles
+        maxcycles = cfg.getint(sect, 'pseudo-mgcycles-max')
+        mincycles = cfg.getint(sect, 'pseudo-mgcycles-min')
+
+        if maxcycles < mincycles:
+            raise ValueError('The maximum number of multigrid cycles must '
+                             'be greater than or equal to the minimum')
+
+        level = 0
+
+        # Generate multiple cfgs for the multigrid systems
+        mgcfg = [Inifile(cfg.tostr()) for i in range(nlevels)]
+        for deg in range(1, order + 1):
+            mgcfg[deg - 1].set('system', 'order', order - (deg - 1))
+            for sec in cfg.sections():
+                m = re.match(r'solver-(.*)-mg-p{0}'.format(deg), sec)
+                if m:
+                    mgcfg[order - deg].rename_section(m.group(0),
+                                                      'solver-' + m.group(1))
+
+        # Patch idxcurr
+        def gidxcurr(self): return self._idxcurr[self.level]
+        def sidxcurr(self, y): self._idxcurr[self.level] = y
+        idxcurr = property(gidxcurr, sidxcurr)
+
+        # Patch regs
+        def gregs(self): return self._regs[self.level]
+        def sregs(self, y): self._regs[self.level] = y
+        regs = property(gregs, sregs)
+
+        # Patch system
+        def gsystem(self): return self._system[self.level]
+        def ssystem(self, y): self._system = y
+        system = property(gsystem, ssystem)
+
+        def _get_reg_banks(self, nreg):
+            regs, _regs, regidx = [], [], list(range(nreg))
+            # Create a proxylist of matrix-banks for each storage register
+            for l in range(nlevels):
+                for i in regidx:
+                    _regs.append(
+                        proxylist([self.backend.matrix_bank(em, i)
+                                   for em in self._system[l].ele_banks])
+                    )
+                regs.append(_regs)
+
+            return regs, regidx, [0]*nlevels
+
+        @staticmethod
+        def _get_system(systemcls, *args, **kwargs):
```

Comment:
Not quite sure what do you mean since I want to pass the multigrid cfgs.

Because
class BaseSystem:
def init(self, backend, rallocs, mesh, initsoln, nreg, cfg):,
shouldn't we have
def _get_system(systemcls, *args):
return [systemcls(*args[:-1], mgcfg[i]) for i in range(nlevels)] .